Frontend/Gmfctn/src/app/core/services/achievement-service/achievement-service.service.ts

```diff
@@ -0,0 +1,54 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable, OnDestroy } from '@angular/core';
+import { BehaviorSubject, EMPTY, Observable, of } from 'rxjs';
+import { catchError, count, map, tap } from 'rxjs/operators';
+import { Achievement } from 'src/app/shared/models/achievement';
+import { apiUrl } from 'src/environments/environment';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AchievementServiceService implements OnDestroy {
+
+  achievements$: BehaviorSubject<Achievement[]> = new BehaviorSubject(null as unknown as Achievement[]);
+
+  constructor(private httpClient: HttpClient) { }
+
+  ngOnDestroy(): void {
+    this.achievements$.complete();
+  }
+
+  getAllAchievements(): Observable<Achievement[]> {
+    return this.httpClient.get<Achievement[]>(`${apiUrl}api/achievement`)
+    .pipe(tap(achievements => {
+      this.achievements$.next(achievements);
+    }));
+  }
+
+  sendRequest(body: object): Observable<object> {
+    return this.httpClient.post(`${apiUrl}api/request/request`, body)
+    .pipe();
+  }
+
+  getUserAchievementsById(id: string): Observable<Achievement[]> {
+    return this.httpClient.get<Achievement[]>(`${apiUrl}api/achievement/get-user-achievements?UserId=${id}`)
+    .pipe(map(achievements => this.loadAchievementsIcons(achievements)));
+  }
+
+  loadAchievementsIcons(achievements: Achievement[]): Achievement[] {
+    const achievementsWithIcon = achievements;
```

Comment:
this line makes no sense. If your goal was to achieve immutability then you'd need to do [...achievements], but you're mutating achievement.iconId a bit later so if the immutability was a goal then pay attention to that line also. My personal opinion is to write immutable functions as it's much easier to debug

Reply:
Fixed